processor/resourcedetectionprocessor/internal/env/env.go

```diff
@@ -0,0 +1,77 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package env provides a detector that loads resource information from
+// the OTEL_RESOURCE environment variable. A list of labels of the form
+// `<key1>=<value1>,<key2>=<value2>,...` is accepted. Domain names and
+// paths are accepted as label keys.
+package env
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+const TypeStr = "env"
+
+// Environment variable used by "env" to decode a resource.
+const envVar = "OTEL_RESOURCE"
+
+type Detector struct{}
+
+func (d *Detector) Detect(context.Context) (pdata.Resource, error) {
+	res := pdata.NewResource()
+	res.InitEmpty()
+
+	labels := strings.TrimSpace(os.Getenv(envVar))
+	if labels == "" {
+		return res, nil
+	}
+
+	err := initializeAttributeMap(res.Attributes(), labels)
+	if err != nil {
+		res.Attributes().InitEmptyWithCapacity(0)
+		return res, err
+	}
+
+	return res, nil
+}
+
+var labelRegex = regexp.MustCompile(`^\s*([[:ascii:]]{1,256}?)\s*=\s*([[:ascii:]]{0,256}?)\s*(?:,|$)`)
+
+func initializeAttributeMap(am pdata.AttributeMap, s string) error {
+	for len(s) > 0 {
+		match := labelRegex.FindStringSubmatch(s)
+		if len(match) == 0 {
+			return fmt.Errorf("invalid resource format, remainder: %s", s)
+		}
+
+		value := match[2]
+		var err error
+		if value, err = url.QueryUnescape(value); err != nil {
```

Comment:
I didn't quite follow from the spec, is the purpose of the query encoding so that = and , are being used as special separator values and the encoding allows those values being used in the actual resource key/values not to be confused with the separators?

Reply:
I was going off this [PR](https://github.com/open-telemetry/opentelemetry-specification/pull/505) / [comment](https://github.com/open-telemetry/opentelemetry-specification/pull/505/files#r389162281).

I believe the intention was to simplify the definition of how to encode resources in an environment variable by borrowing (i.e. referencing) the spec that is already defined for correlation context, since these are basically the same format. The reason why we need this isn't called out particularly well in that PR, but I assume it's for the reason you mentioned.

Even though that PR probably won't get merged, there seemed to be support for using this definition for resources supplied in environment variables.